Add JavaScript unit tests and refactor

sirbrillig opened the pull request:
Add basic JavaScript unit tests for `wp.widgets.Form`. They can be run using `npm test` on the command-line (you may need to run `npm install` first to get the tests properly set up).

This also makes some changes to the `wp.widgets.Form` class to make the code more concise and more testable. There are also a few bugs that are fixed. Changes include:

- Explicitly pass underscore to the IIFE
- Remove unnecessary use of `_.clone`
- Export `Form` as a module when within CommonJS
- Update `Form.initialize` to deep merge config with defaults
- Remove merging `default_instance` with the model in all places except `Form.getValue`
- Remove adding missing `title` property to model in `Form.sanitize`
- Simplify the `Form.model.validate` method
- Simplify the `Form.initialize` constructor
- Add `Form.renderNotificationsToContainer` to allow testing `Form.renderNotifications`
- Always remove all sanitize-added notifications in `Form.model.validate`

This needs to be tested with a real-world Form to be sure that the above changes do not break anything.

There's also more simplifications that could be made, but I wanted to get a PR up for the purposes of discussion.

sirbrillig committed:
Add initial basic test

sirbrillig committed:
Alter extendWidgetControl to a (mostly) pure function

Changes the function to accept an object, alter that object's prototype,
and return it. The caller is then responsible for passing the
constructor function to change.

This allows the function to be more easily tested since it does not
directly mutate global variables (although of course changing the
prototype of a global variable will have a a global effect).

Also adds `wp` and `_` as parameters to the IIFE to make their usage
explicit.

sirbrillig committed:
Add truthy test for isJsWidgetControl

Not currently using real Controls, but it's something.

sirbrillig committed:
Fix missing jsdoc rules to satisfy eslint

sirbrillig committed:
Explicitly pass underscore to the IIFE

IIFEs are good tools for encapsulating functionality and to maintain
that encapsulation, all inputs and outputs (the exported API) need to be
explicit. To simply alias variables, variable assignment is a less
cumbersome tool.

sirbrillig committed:
Remove unnecessary use of `_.clone`

`_.clone()` only performs a shallow copy and `_.extend()` is already
doing that here.

sirbrillig committed:
Export Form as a module when within CommonJS

sirbrillig committed:
Add initial tests for Form

sirbrillig committed:
Add customize-base for tests

sirbrillig committed:
Update Form.initialize to deep merge config with defaults

sirbrillig committed:
Update tests to prove setState does not unexpectedly mutate the model

sirbrillig committed:
Add tests for Form.model.validate()

sirbrillig committed:
Add additional test to verify merging properties for setState()

sirbrillig committed:
Remove test for sanitize adding a title property

sirbrillig committed:
Remove merging default_instance in validate

sirbrillig committed:
Remove merging default_instance in setState()

sirbrillig committed:
Add jsdom to use real jQuery for tests

sirbrillig committed:
Add test for two-way binding form and model data

sirbrillig committed:
Simplify the form.model.validate method

sirbrillig committed:
Simplify Form constructor

sirbrillig committed:
Remove merging default_instance in sanitize

sirbrillig committed:
Remove adding missing title property to model in sanitize

This does not seem necessary.

sirbrillig committed:
Add Form.renderNotificationsToContainer to allow testing

Debounced functions are async by nature and harder to test. This
rewrites `renderNotifications` as a debounced proxy for
`renderNotificationsToContainer`.

sirbrillig committed:
Add tests for renderNotificationsToContainer

sirbrillig committed:
Update validate tests to prove notifications are always removed

sirbrillig committed:
Always remove all sanitize-added notifications in validate

sirbrillig committed:
Add missing semicolon to validate function

sirbrillig committed:
Ignore js tests in jshintignore

sirbrillig committed:
Add `npm test` to travis config

sirbrillig committed:
Remove trailing comma to placate jshint

westonruter committed:
Add root slash for .jshintignore

sirbrillig committed:
Add tests dir to jscsrc ignore

sirbrillig committed:
Add tests/js/lib to eslintignore

westonruter commented on .travis.yml:
```diff
@@ -18,9 +18,11 @@ install:
     - export DEV_LIB_PATH=dev-lib
     - if [ ! -e "$DEV_LIB_PATH" ] && [ -L .travis.yml ]; then export DEV_LIB_PATH=$( dirname $( readlink .travis.yml ) ); fi
     - if [ ! -e "$DEV_LIB_PATH" ]; then git clone https://github.com/xwp/wp-dev-lib.git $DEV_LIB_PATH; fi
+    - npm install
```
I believe `npm install` is already being run by `travis.install.sh`, per https://github.com/xwp/wp-dev-lib/blob/d478bb3bc2837f8a3a5019aebf8d6bc8b9f53c38/check-diff.sh#L377-L380

sirbrillig replied:
Oh! Thanks, I didn't know about wp-dev-lib. I clearly didn't set up my fork of this repo correctly with the submodules.

sirbrillig committed:
Remove duplicate "npm install"

sirbrillig committed:
Simplify renderNotificationsToContainer

westonruter commented on js/widget-form.js:
```diff
@@ -10,9 +10,125 @@ if ( ! wp.widgets.formConstructor ) {
+		var propertiesConfig = properties ? properties.config : {};
+		args.config = _.extend( {}, defaultConfig, config, propertiesConfig || {} );
+		return args;
+	}
```
Should these functions get jsdoc?

sirbrillig replied:
That's a good point. They are "private" functions and cannot be used by subclasses directly, so at some level there's no need to document their API. But of course jsdoc is not just for documenting the public API, it can also be useful for anyone editing this file in the future. I think it would be a good idea.

sirbrillig replied:
Added in bafd1b9. Also, I realize these private functions kind of clutter up the top of the file. I just put them there for convenience. They could just as easily be in their own "namespace" object and/or at the bottom of the file and/or in a separate file. I'm very open to other thoughts about that.

They could even be function properties on the `Form` object, but I'd suggest against that unless there's a reason they would be called directly or overridden (making them private is currently no different than having their functionality inside of the current public functions).

sirbrillig commented:
@westonruter Another question I had, regarding the code under test: Is it a good idea to allow adding arbitrary properties to the object via the constructor argument? It seems a little bit like mixing two inheritance mechanisms in one place: the `Form` can be subclassed, but also you can override its properties and methods using the object passed in the constructor like a mixin.

```js
var form = new Form({ 
    foo: 'bar', 
    sanitize: function( val ) { return _.extend({}, val, {hello: 'world'}); } 
} );
form.foo; // 'bar'
form.sanitize( {} ); // {hello: 'world'}
```

sirbrillig committed:
Add jsdoc to each helper method

sirbrillig committed:
Rename renderNotificationsTemplate to renderMarkupToContainer

The previous helper had too many arguments

westonruter commented:
@sirbrillig good question. It's probably not good form to do that. What do you suggest? Mark certain properties as being protected and throw `Error` if they are attempted to be overridden? Or some other refactoring of `Form`?

sirbrillig commented:
@westonruter hm... I guess my first question is: what is the "normal" or "expected" use-case of that feature? That is, what are the intended arguments to the constructor?

I'll make a guess (but I could definitely be missing something): it's intended to allow setting/overriding `id_base`, `model`, `container`, and (certain properties of) `config`. If that's true, then I would say to just copy over those properties explicitly, and silently ignore any other properties in the parameter object.

I could see the benefit of throwing if there's an unexpected property, in that it might make it easier to find bugs or typos, but I could also imagine a situation where the argument object is constructed by something else and passed into the constructor programmatically and we might want to just ignore unexpected properties. Also, the constructor _does_ already throw if the required properties are not present.

So I could go either way, but my instinct would be to ignore unexpected properties and explicitly assign expected ones. What do you think?

westonruter commented:
> So I could go either way, but my instinct would be to ignore unexpected properties and explicitly assign expected ones. What do you think?

@sirbrillig I think you're right. I think I initially was allowing anything to be passed in because it wasn't used enough to see clearly what the usage patterns should be. Really, the only params that absolutely would need to be passed in which are unique to each instance would be `model` and `container`. The `id_base`, `default_instance`, and other config parameters really should be defined on the subclass `prototype` itself.

So then for PHP should export its config _once_. And actually, this is _already_ the case for the core widgets that are implemented: https://github.com/xwp/wp-js-widgets/blob/38d8298528737add34594efdb4db8d649355a140/php/class-wp-adapter-js-widget.php#L85-L89

And looking at the existing usages for admin screen and customizer pane, only the `model` and `container` properties are being used:
* https://github.com/xwp/wp-js-widgets/blob/910bcc9cbff1a1c6cee8656b2324b9d2e9766494/js/admin-js-widgets.js#L52-L56
* https://github.com/xwp/wp-js-widgets/blob/910bcc9cbff1a1c6cee8656b2324b9d2e9766494/js/customize-js-widgets.js#L173-L176

So yeah, let's update `Form` to only honor passing a `model` and `container` args, and we can also let `Form` be marked as `@abstract` to then also throw errors if `initialize` is called and `this` doesn't have `id_base` or `default_config` inherited from the prototype.

Does that make sense?

sirbrillig committed:
Change tests so only model and container can be Form arguments

sirbrillig committed:
Explicitly pluck Form arguments when assigning

Only allows `model` and `container`, ignoring all other properties.

sirbrillig committed:
Change tests so Form constructor requires id_base and default_instance

sirbrillig committed:
Require id_base and config.default_instance in Form constructor

sirbrillig committed:
Add test for memory leak in Form.model.validate

sirbrillig commented:
@westonruter that makes sense. I added tests to that effect and then changed the constructor. 

Another thing I just noticed (which might not actually be relevant) is that because `Form` overrides its `model.validate` method with a closure that includes methods on the `Form`, the model will keep a reference to that `Form` even after the Form is destroyed. I don't know where the model comes from or if it lives beyond the Form's existence, but maybe we should restore the original validate method in `destruct`? 

Alternatively, if this is a concern, instead of overriding the `model.validate` method, we could just run our own validation in the form change event handler before the data is sent to the model. Fewer mutations of external objects would reduce the chance for errors to creep in. That is, unless it's expected that the model might receive data from another source which would require the modified validation.

I added a test to show the memory leak in case we want to change anything later.

sirbrillig committed:
Move Form private functions to bottom of file

It's easier to read the purpose of the file if the class comes first.

sirbrillig committed:
Add test for changing notification area height on show

sirbrillig committed:
Fix changing height of Form notifications area

westonruter commented:
@sirbrillig:

> Another thing I just noticed (which might not actually be relevant) is that because Form overrides its model.validate method with a closure that includes methods on the Form, the model will keep a reference to that Form even after the Form is destroyed. I don't know where the model comes from or if it lives beyond the Form's existence, but maybe we should restore the original validate method in destruct?

Oh yeah, that's a good point! Yeah, the `model` can indeed live beyond the `Form`'s existence and so restoring the original `validate` method should indeed be done in the form is destructed.

But actually, it would probably be even better to create a new `Value` instance inside the `Form`'s constructor and then to `sync` this new value with the supplied model. This new `Value` instance could then have its own unique `validate` method, and when the `Form` is destructed it can just be `unsync`'ed and destroyed. 

So inside of `initialize`:

```js
form.model = new api.Value();
form.model.validate = function( value ) {
	// Current logic in our overridden validate method.
	return value;
};
form.model.set( args.model.get );
form._externalModel = args.model;
form.model.sync( form._externalModel );
```

And then in `destruct`:

```js
form.model.unsync( form._externalModel );
form.model = null;
```

This approach will keep the original `args.model` object pristine and it will also prevent a memory leak.

> Alternatively, if this is a concern, instead of overriding the model.validate method, we could just run our own validation in the form change event handler before the data is sent to the model. Fewer mutations of external objects would reduce the chance for errors to creep in. That is, unless it's expected that the model might receive data from another source which would require the modified validation.

In this case it would just mean that the form should be forbidden from directly accessing `this.model`, right? Subclasses would be restricted to only using `form.getValue()`. This would make sense to me actually. And then the `form.setState()` method would then be what implements the `validate` logic? The `model` shouldn't be expected to be modified from another place outside of the `Form`, or if it is, we can assume that it is already validated (for example, populating the customizer settings with a previous revision).

sirbrillig commented:
> In this case it would just mean that the form should be forbidden from directly accessing this.model, right? Subclasses would be restricted to only using form.getValue().

Right. I think I like this approach the most because it's just less complexity (and one less thing for the subclass to "know about"), even though the model duplication/sync/unsync approach would certainly work.

westonruter commented:
@sirbrillig ok, sounds good! You want to incorporate that into this PR?

sirbrillig commented:
@westonruter I'm fine to do that unless you'd like to keep the change separate?

sirbrillig committed:
Change tests so that validate becomes a method on Form

This replaces overriding the `model.validate` method by requiring using
`form.setState()` to update the model

sirbrillig committed:
Move model.validate to Form.validate

Form.validate is now called by Form.setState. This avoids mutating the model.

sirbrillig committed:
Make model.get always explicit

sirbrillig committed:
Add tests to be sure two-way binding uses validate

sirbrillig committed:
Add placeholder for using setState in linkPropertyElements

sirbrillig committed:
Add tests for destruct and field references

sirbrillig commented:
As part of the changes I simplified two-way data binding a bit (see 16f9762). I'm not sure if the resulting simplified version covers everything that the `sync` version did. The primary difference is that we're only listening to the `change` event, rather than `propertychange`, `keyup`, and `input`. There's lots of other ways to sync the data so if this doesn't work we can try something else.

westonruter commented:
@sirbrillig oh, I didn't consider how `createSyncedPropertyValue` related here. A problem with `change` is that it will only trigger when the user has blurred the field, resulting in a poorer preview experience. This is what `wp.customize.Element` is supposed to abstract away, using the appropriate DOM events to listen for changes to the underlying `input`.

I think that `createSyncedPropertyValue` needs to be restored, but with a change to how `sync` is done. Inside of `propertyChangeListener` we can do this change instead:

```diff
--- a/js/widget-form.js
+++ b/js/widget-form.js
@@ -272,15 +272,15 @@ wp.widgets.Form = (function( api, $ ) {
 		 * @returns {object} Property value instance.
 		 */
 		createSyncedPropertyValue: function createSyncedPropertyValue( root, property ) {
-			var propertyValue, rootChangeListener, propertyChangeListener;
+			var form = this, propertyValue, rootChangeListener, propertyChangeListener;
 
 			propertyValue = new api.Value( root.get()[ property ] );
 
 			// Sync changes to the property back to the root value.
 			propertyChangeListener = function( newPropertyValue ) {
-				var rootValue = _.clone( root.get() );
-				rootValue[ property ] = newPropertyValue;
-				root.set( rootValue );
+				var newState = {};
+				newState[ property ] = newPropertyValue;
+				form.setState( newState );
 			};
 			propertyValue.bind( propertyChangeListener );
```

This will ensure that the `model` is not directly mutated so that our `validate` logic will still apply. 

Does that make sense?

sirbrillig committed:
Modify createSyncedPropertyValue to use setState

This makes sure that data from the DOM is validated before changing the
model.

sirbrillig commented:
Ah, nice. Don't know why I didn't see that part of the function before. I guess I wasn't clear what was going on in there. I made the changes you suggested and the tests all pass. 👍 (I rebased away the old version.)

One last question:

In `renderNotificationsToContainer`, this command is run to change the CSS of the notifications container after it is shown: `container.css( 'height', 'auto' );`. I'm curious why that's only done when the container is shown and it's not changed when it's hidden? Could the height just be part of the actual CSS for the Form instead of being set with javascript? I guess that also makes me wonder: why not do the animation with CSS instead of jQuery?

westonruter commented:
> Ah, nice. Don't know why I didn't see that part of the function before. I guess I wasn't clear what was going on in there. I made the changes you suggested and the tests all pass.

Great!

> I'm curious why that's only done when the container is shown and it's not changed when it's hidden? Could the height just be part of the actual CSS for the Form instead of being set with javascript? I guess that also makes me wonder: why not do the animation with CSS instead of jQuery?

@sirbrillig Humm, I don't have a good answer for you. I can't remember why `height:auto` was added 😦 If you see no good reason for it, the I'd say to go ahead and remove. The styling/animation logic here is copied from `wp.customize.Control#renderNotifications`. If you have a CSS solution then more power to you! But I think that CSS transitions/animations weren't used because they are not supported by IE<11.

sirbrillig commented:
TIL that CSS transitions aren't available in IE < 10. It's amazing how quickly I've come to think of them as natural. Well, I don't know why `height: auto` is in there either, but it's a relatively minor thing either way. I'd be inclined to just remove it and re-add it if/when it becomes apparent that there's a problem that needs to be solved.

westonruter commented:
Great! I'll give this PR a test soon and see if there's any issues on my end and then merge. Excellent work here. Really appreciate the depth you went and the insights you have.

sirbrillig committed:
Remove fixed todo

sirbrillig committed:
Fix `properties.container` jsdoc for Form.initialize

westonruter committed:
Replace deprecated name param with field param

westonruter committed:
Ensure that id_base is defined on widget form prototypes

westonruter committed:
Merge props to validate on top of existing model

westonruter commented:
@sirbrillig I found that the widgets weren't getting `id_base` defined on the widget prototypes, so I added 5d7ccb0. This means a widget should be defined via:

```js
wp.widgets.formConstructor.foo = wp.widgets.Form.extend({
     id_base: 'foo'
});
```

So there is some redundancy there. In reality they should never be different, so it seems a shame to duplicate.

sirbrillig committed:
Update setState test to better prove validate

sirbrillig committed:
Add test for early merge of model properties in setState

sirbrillig committed:
Remove unused variable in setState

sirbrillig commented:
> So there is some redundancy there. In reality they should never be different, so it seems a shame to duplicate.

I'm not sure I understand. Do you mean, the `id_base` property should always be the same as the name of the property or variable to which that object is assigned?

So `const HelloWidget = Form.extend( {} )`  should result in `HelloWidget.id_base === 'HelloWidget'`?

westonruter commented:
> I'm not sure I understand. Do you mean, the id_base property should always be the same as the name of the property or variable to which that object is assigned?

- [x] Per https://github.com/xwp/wp-js-widgets/pull/35#commitcomment-20881119 I think we can kill `id_base` entirely 😄

sirbrillig committed:
Remove id_base from Form tests

sirbrillig committed:
Remove id_base property from Form

sirbrillig committed:
Remove id_base from core-adapter-widgets

sirbrillig commented:
I removed `id_base` from `Form`. Let me know if there's anything else!

westonruter committed:
Rename validateForm to assertValidForm

westonruter committed:
Update readme, bump version, add props

[ci skip]
